File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/modules/CustomHttpHeaders.kt
package com.mapbox.rctmgl.modules

import com.mapbox.common.*

object CustomHttpHeaders : HttpServiceInterceptorInterface {
    init {}

    val map = mutableMapOf<String, String>()

    fun addCustomHeader(headerName: String, headerValue: String) {
        HttpServiceFactory.getInstance().setInterceptor(
            this
        )
        map.put(headerName, headerValue)
    }

    fun removeCustomHeader(headerName: String) {
        map.remove(headerName)
    }

    override fun onRequest(request: HttpRequest): HttpRequest {
        for (entry in map.entries.iterator()) {
            request.headers[entry.key] = entry.value
        }
        return request
    }

    override fun onDownload(download: DownloadOptions): DownloadOptions {
        for (entry in map.entries.iterator()) {
            download.request.headers[entry.key] = entry.value
        }
        return download
    }

    override fun onResponse(response: HttpResponse): HttpResponse {
        return response
    }
}

File: ios/RCTMGL-v10/CustomHttpHeaders.swift
import MapboxMaps

class CustomHttpHeaders: HttpServiceInterceptorInterface {
  static var shared: CustomHttpHeaders = {
    let headers = CustomHttpHeaders()
    headers.install()
    return headers
  }()

  var customHeaders: [String: String] = [:]

  func install() {
    HttpServiceFactory.getInstance().setInterceptorForInterceptor(self)
  }

  func reset() {
    HttpServiceFactory.getInstance().setInterceptorForInterceptor(nil)
  }

  // MARK: - HttpServiceInterceptorInterface

  func onRequest(for request: HttpRequest) -> HttpRequest {
    customHeaders.forEach {(key, value) in
      request.headers[key] = value
    }
    return request
  }

  func onDownload(forDownload download: DownloadOptions) -> DownloadOptions {
    customHeaders.forEach {(key, value) in
      download.request.headers[key] = value
    }
    return download
  }

  func onResponse(for response: HttpResponse) -> HttpResponse {
    return response
  }
}
